modules/ztp-image-based-upgrade-procedure.adoc

```diff
@@ -0,0 +1,144 @@
++
+[source,yaml]
+----
+apiVersion: lcm.openshift.io/v1alpha1
+kind: ImageBasedGroupUpgrade
+metadata:
+  name: example-upgrade
+  namespace: default
+spec:
+  clusterLabelSelectors: <1>
+  - matchExpressions:
+    - key: name
+      operator: In
+      values:
+      - spoke1
+      - spoke2
+      - spoke6
+  ibuSpec:
+    seedImageRef: <2>
+      image: quay.io/seed/image:4.16.0-rc.1
+      version: 4.16.0-rc.1
+      pullSecretRef:
+        name: "<seed_pull_secret>"
+  extraManifests: <3>
+    - name: example-extra-manifests
+      namespace: openshift-lifecycle-agent
+  oadpContent: <4>
+    - name: oadp-cm
+      namespace: openshift-adp
+  plan: <5>
+  - actions:
+    - Prep
+    rolloutStrategy:
+      maxConcurrency: 2
+      timeout: 2400
+  - actions:
+    - Upgrade
+    rolloutStrategy:
+      maxConcurrency: 2
+      timeout: 30
+  - actions:
+    - FinalizeUpgrade
+    rolloutStrategy:
+      maxConcurrency: 10
+      timeout: 3
+----
+<1> Specify the set of clusters that you want to upgrade.
+<2> Defines the target platform version, the seed image to be used, and the secret required to access the image.
+<3> (Optional) Specify the list of `ConfigMap` resources that contain your custom catalog sources to retain after the upgrade, and your extra manifests to apply to the target cluster that are not part of the seed image.
+<4> Specify the list of `ConfigMap` resources that contain the OADP `Backup` and `Restore` CRs.
+<5> Defines the upgrade plan. This example plan shows that all steps have separate rollout strategies.
+// Do we want to include all this? Do we need to add more information?
+
+. Apply the `example-upgrade.yaml` file by running the following command on the hub cluster:
++
+[source,terminal]
+----
+$ oc apply -f example-upgrade.yaml
+----
+
+. Monitor the status updates by running the following command:
++
+--
+[source,terminal]
+----
+$ oc get ibgu -o yaml
+----
+
+[source,yaml]
+----
+[...]
+status:
+  clusters:
+  - completedActions:
+    - action: Prep
+    name: spoke1
+  - completedActions:
+    - action: Prep
+    name: spoke2
+  - failedActions:
+    - action: Prep
+    name: spoke6
+  conditions:
+  - message: All plan steps are completed
+    reason: Completed
+    status: "False"
+    type: Progressing
+[...]
+----
+
+Based on the results of the step, {cgu-operator} adds a label to the clusters.
+In this example, `lcm.openshift.io/ibgu-prep-failed` label is applied to `spoke6` after a failed `Prep` stage transition.
+
+After investigating the reason for failure, you can append an `AbortOnFailure` step to move the cluster back to the `Idle` stage.
+This deletes any resources that are related to the upgrade and allows you to retry the upgrade after troubleshooting.
+--
+
+. (Optional) Append the `AbortOnFailure` action as a separate step to your plan in the `ImageBasedGroupUpgrade` CR:
++
+--
+// Please let me know how you would append this step in practice.
```

Comment:
you can add items to plan list
```
plan:
- actions: ["Prep"]
```
after oc patch
```
plan:
- actions: ["Prep"]
- actions: ["AbortOnFailure"]
```

you can not modify the existing items on the plan or change the current order.
so for this example
```
  plan: <5>
  - actions:
    - Prep
    rolloutStrategy:
      maxConcurrency: 2
      timeout: 2400
  - actions:
    - Upgrade
    rolloutStrategy:
      maxConcurrency: 2
      timeout: 30
  - actions:
    - FinalizeUpgrade
    rolloutStrategy:
      maxConcurrency: 10
      timeout: 3
```
you can't add AbortOnFailure on this, but you can have AbortOnFailure from the beginning and it will automatically aborts the clusters that fail on prep or upgrade:
```
    plan:                     
    - actions:        
      - Prep   
      rolloutStrategy:                                                        
        maxConcurrency: 200                                                   
        timeout: 15    
    - actions:       
      - AbortOnFailure 
      rolloutStrategy:   
        maxConcurrency: 200
        timeout: 10
    - actions:       
      - Upgrade   
      rolloutStrategy:
        maxConcurrency: 200
        timeout: 60
    - actions:
      - AbortOnFailure
      rolloutStrategy:
        maxConcurrency: 200
        timeout: 10
    - actions:
      - FinalizeUpgrade
      rolloutStrategy:
        maxConcurrency: 200
        timeout: 30
```

Reply:
@sudomakeinstall2 Thank you! I have a few follow-up questions:
1) "you can not modify the existing items on the plan or change the current order." - this means that users can't modify a plan that has already started, so after applying the IBGU?
2) If we include AbortOnFailure actions from the start, every cluster will be moved back to Idle stage when the `lcm.openshift.io/ibgu-<action>-failed` label is added? If so, how can the user troubleshoot in this case?

Reply:
1- No you can't modify the existing entries in the plan after the IBGU is created. You can only add new entries to the plan list.
2- Yes troubleshooting in this case will be hard, since there will be only a message in the ibgu status that indicates Prep or Upgrade has failed. So it is better to not have all the actions from the start and add them step by step. So if one step fails you can troubleshoot it. for example create ibgu with only `prep`. Then if some clusters failed prep, after troubleshooting them add `abortOnFailure` to abort failed ones and then add `upgrade` to upgrade the successful ones and so on.

Also there is a separate `Abort` or `Rollback` scenario that is needed to be added to docs. In cases where the user wants to `Abort` or `Rollback` some clusters regardless of failures. They can create a separate IBGU and put `Abort` or `Rollback` in actions. This way user can abort clusters that have successfully finished prep, or rollback successful upgrades.